controllers/provisioning_controller.go

```diff
@@ -335,6 +317,68 @@ func (r *ProvisioningReconciler) provisioningInfo(provConfig *metal3iov1alpha1.P
 	}
 }
 
+//Ensure Finalizer is present on the Provisioning CR when not deleted and
+//delete resources and remove Finalizer when it is
+func (r *ProvisioningReconciler) checkForCRDeletion(info *provisioning.ProvisioningInfo) (bool, error) {
+	// examine DeletionTimestamp to determine if object is under deletion
+	if info.ProvConfig.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object. This is equivalent
+		// to registering our finalizer.
+		if !containsString(info.ProvConfig.ObjectMeta.Finalizers,
+			metal3iov1alpha1.ProvisioningFinalizer) {
+			// Add finalizer becasue it doesn't already exist
+			info.ProvConfig.ObjectMeta.Finalizers = append(info.ProvConfig.ObjectMeta.Finalizers,
+				metal3iov1alpha1.ProvisioningFinalizer)
+			if err := r.Client.Update(context.Background(), info.ProvConfig); err != nil {
+				return false, errors.Wrap(err, "failed to update Provisioning CR with its finalizer")
+			}
+		}
+		return false, nil
+	} else {
+		// The Provisioning object is being deleted
+		if containsString(info.ProvConfig.ObjectMeta.Finalizers, metal3iov1alpha1.ProvisioningFinalizer) {
+			err := r.deleteMetal3Resources(info)
+			if err != nil {
+				return false, errors.Wrap(err, "failed to delete metal3 resource")
+			}
+			// Remove our finalizer from the list and update it.
+			info.ProvConfig.ObjectMeta.Finalizers = removeString(info.ProvConfig.ObjectMeta.Finalizers,
+				metal3iov1alpha1.ProvisioningFinalizer)
+			if err = r.Client.Update(context.Background(), info.ProvConfig); err != nil {
+				return true, errors.Wrap(err, "failed to remove finalizer from Provisioning CR")
+			}
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+//Delete Secrets and the Metal3 Deployment objects
+func (r *ProvisioningReconciler) deleteMetal3Resources(info *provisioning.ProvisioningInfo) error {
+	failedSecrets := provisioning.DeleteAllSecrets(info)
+	if len(failedSecrets) > 0 {
+		for _, secret := range failedSecrets {
+			r.Log.V(1).Info("failed to delete secret : ", "secret", secret)
+		}
+		r.Log.V(1).Info("proceeeding with deletion of other resources")
+	}
+
+	err := provisioning.DeleteMetal3Deployment(info)
+	if err != nil {
+		return errors.Wrap(err, "failed to delete metal3 deployment")
+	}
+	err = provisioning.DeleteMetal3StateService(info)
+	if err != nil {
+		return errors.Wrap(err, "failed to delete metal3 service")
+	}
+	err = provisioning.DeleteImageCache(info)
```

Comment:
What do you think about this? We have a lot of repeated functions that just check for notFound
```golang
err = provisioning.IgnoreNotFound(info.Client.AppsV1().DaemonSets(info.Namespace).Delete(context.Background(), imageCacheService, metav1.DeleteOptions{}))
```

```
func IgnoreNotFound(err error) error {
	if apierrors.IsNotFound(err) {
		return nil
	}
	return err
}
```

Reply:
Not really going to help the Reconcile method a whole lot. We currently call apierrors.IsNotFound() twice and the number of return values are different.
In other places in the provisioning package, we use the return value of apierrors.IsNotFound() to decide if we need to create something and not return nil.